lib/matplotlib/dates.py

```diff
@@ -740,12 +746,246 @@ def __call__(self, x, pos=0):
+        ticks = self._locator()
+        if pos is not None:
+            if not np.array_equal(ticks, self._oldticks):
+
+                offset_fmt = ''
+                fmt = self.defaultfmt
+                self._formatter = DateFormatter(fmt, self._tz)
+                tickdate = np.array([num2date(tick).timetuple()[:6]
+                                     for tick in ticks])
+                tickdatetime = [num2date(tick) for tick in ticks]
+
+                # basic algorithm:
+                # 1) only display a part of the date if it changes over the
+                #    ticks.
+                # 2) don't display the smaller part of the date if:
+                #    it is always the same or if it is the start of the
+                #    year, month, day etc.
+                # fmt for most ticks at this level
+                fmts = self.formats
+                # format beginnings of days, months, years, etc...
+                zerofmts = self.zero_formats
+                # offset fmt are for the offset in the upper left of the
+                # or lower right of the axis.
+                offsetfmts = self.offset_formats
+
+                # determine the level we will label at:
+                # mostly 0: years,  1: months,  2: days,
+                # 3: hours, 4: minutes, 5: seconds, 6: microseconds
+                for level in range(5, -1, -1):
+                    if len(np.unique(tickdate[:, level])) > 1:
+                        break
+
+                # level is the basic level we will label at.
+                # now loop through and decide the actual ticklabels
+                zerovals = [0, 1, 1, 0, 0, 0, 0]
+                ticknew = []
+                zeroed = False
+                for nn in range(len(tickdate)):
+                    if level < 5:
+                        if tickdate[nn][level] == zerovals[level]:
+                            fmt = zerofmts[level]
+                            zeroed = True
+                        else:
+                            fmt = fmts[level]
+                    else:
+                        # special handling for sconds + microseconds
+                        if (tickdatetime[nn].second == 0 and
+                                tickdatetime[nn].microsecond == 0):
+                            fmt = zerofmts[level]
+                            zeroed = True
+                        else:
+                            fmt = fmts[level]
+
+                    newtickla = tickdatetime[nn].strftime(fmt)
+                    ticknew.append(newtickla)
+
+                # special handling of seconds and microseconds:
+                # strip extra zeros and decimal if possible...
+                # this is complicated by two factors.  1) we have some
+                # level-4 strings here (i.e. 03:00, '0.50000', '1.000')
+                # 2) we would like to have the same number of decimals for
+                # each string (i.e. 0.5 and 1.0).
+                done = False
```

Comment:
looks like the whole block (from done = False to line 961) can be replaced by
```
                trailing_zeros = min(
                    (len(s) - len(s.rstrip('0')) for s in ticknew if '.' in s),
                    default=None)
                if trailing_zeros:
                    for nn in range(len(ticknew)):
                        if '.' in ticknew[nn]:
                            ticknew[nn] = \
                                ticknew[nn][:-trailing_zeros].rstrip('.')
```
(at least this passes the test...)

Reply:
Thats much better.  Thanks!  I still don't instinctively do list comprehensions....

I left the check for the level in, because there is not much point in doing the loop if we aren't at level 5.  Sure its not that expensive, but...